rcl/include/rcl/graph.h

```diff
@@ -524,6 +525,122 @@ rcl_count_subscribers(
   const char * topic_name,
   size_t * count);
 
+/// Returns a list of all publishers to a topic.
+/// Each element in the list will contain the node name, node namespace, topic type,
+/// gid and the qos profile of the publisher.
+/**
+ * The `node` parameter must point to a valid node.
+ *
+ * The `topic_name` parameter must not be `NULL`.
+ *
+ * The `no_mangle` parameter determines if the provided topic_name should be
+ * expanded to its fully qualified name.
```

Comment:
Is this accurate? For me the mangle has to do with the "ROS topic name" versus the underlying "middleware topic name" which may or may not be the same. It doesn't imply anything about relative versus absolute (or fully qualified) names.

Reply:
Yes, I think the wording of the comment is incorrect.

Should the topic name expansion be done in `rcl` or in `rclcpp` and `rclpy`?
See my comment [here](https://github.com/ros2/rclcpp/pull/960#discussion_r364215929).
I also don't find consistency on how this is done in different functions, e.g.:
- `rcl_publisher_init` [does the topic name expansion](https://github.com/ros2/rcl/blob/024c3ee50d9b79a85665efe1c191cc1225026519/rcl/src/rcl/publisher.c#L74-L156).
- `rcl_count_publishers` [does not the expansion]( https://github.com/ros2/rcl/blob/024c3ee50d9b79a85665efe1c191cc1225026519/rcl/src/rcl/graph.c#L374).

In general, the expansion is done in `rcl`, I think that the count publishers/subscriptions functions are the lonely exceptions. We should also create a function in `rcl` that does the hole expand and remap thing, to avoid repeating ourselves (and repeating that in [upper layers](https://github.com/ros2/rclcpp/blob/33a755c535d654c97401eb6d199580368c19eb40/rclcpp/src/rclcpp/expand_topic_or_service_name.cpp)).

Reply:
I think that it may depend on how it is used. For `rcl_count_publishers()` it may be possible that the caller passes an already mangle topic name? I dunno without digging in some. All I'm saying is that it may intentionally be that way.

Reply:
Updated the docstring.

Reply:
@mm318 I don't see the comment updated.

> I think that it may depend on how it is used. For rcl_count_publishers() it may be possible that the caller passes an already mangle topic name? I dunno without digging in some. All I'm saying is that it may intentionally be that way.

Yeah, maybe doing remapping and expansion is not compatible with passing an already mangled name.
But in the case of `no_mangle=false`, it makes sense to do the `remapping/expansion` in rcl, to avoid repeating code like [this](https://github.com/ros2/rclcpp/blob/4a5eed968ca6fa4979214db8ac945fb163919326/rclcpp/src/rclcpp/node_interfaces/node_graph.cpp#L230) (which is probably repeated in `rclpy` too).
And also, it's strange that the `rcl` function consider the `no_mangle` case and the `rclcpp` doesn't.

Anyways, we can continue the discussion after merging this, and correct `rcl_count_*` and these new functions both together. For now, just keep consistensy with `rcl_count_*`.

The comment may be like:
The `no_mangle` parameter determines if the provided topic_name should be mangled with the ROS conventions (`false`), or if it's a valid topic name for the middleware (useful when combining ROS with native middleware (e.g. DDS) apps) (`true`).

Also, add a comment in the docstring stating that the expected topic name should always be fully qualified (regardless of `no_mangle` param).

Reply:
Oops, I updated the other instance of this comment, but forgot to update this one. Thanks, I'll update both instances of the comment with your suggestion.